### PR TITLE
exercises 01 02 03 done

### PR DESCRIPTION
**Pull request by nizamimursudlu**

### CODE REVIEW

File: Week2/homework/exercise01.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// Have the getAnonName function return a new Promise that uses the firstName parameter
+// If the Promise resolves, pass the full name as an argument to resolve with
+// If the Promise rejects, pass an error as the argument to reject with: "You didn't pass in a first name!"
+
+function getAnonName(firstName) {
+  return new Promise((resolve, reject) => {
+    setTimeout(() => {
+      if (firstName) {
+        const fullName = `${firstName} Doe`;
+        resolve(console.log(fullName))
```
Comment (jasperkennis) [inline Week2/homework/exercise01.js:26]: Instead of passing `console.log` into resolve, try to just pass the `fullName`.

File: Week2/homework/exercise01.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      } else {
+        reject(console.log("You didn't pass in a first name!"))
+      }
+    }, 2000)
+  })
+}
+getAnonName("John")
```
Comment (jasperkennis) [inline Week2/homework/exercise01.js:33]: This works, but you are not using the return value from the function. Try using `.then()` and `.catch()` on `getAnonName("John")` and call `console.log` inside those.

File: Week2/homework/exersice02.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// Returns a new Promise
+// If the number is bigger than 10, resolve with the string: "The number is bigger than 10!"
+// If the number is smaller than 10, reject with the error: "Error! The number is smaller than 10..."
+
+
+
+function checkDoubleDigits(number) {
+  return new Promise((resolve, reject) => {
+    if (number > 10) {
+      resolve(console.log("The number is bigger than 10!"))
```
Comment (jasperkennis) [inline Week2/homework/exersice02.js:15]: For this exercise, the same applies as in the previous exercise; try to resolve with the result value, and use `.then` and `.catch`.

File: Week2/homework/pokemon-app/script.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  const select = document.getElementById("select");
+  const url = "https://pokeapi.co/api/v2/pokemon/?limit=151"
+  const button = document.getElementById("button");
+  const div = document.getElementById("div")
+  img = document.createElement("img")
+
+
+  button.addEventListener('click',
+    function fetchData() {
+      fetch(url)
```
Comment (jasperkennis) [inline Week2/homework/pokemon-app/script.js:31]: Wow, I never used this function, I didn't even know it existed, thanks for showing me!

File: Week2/homework/pokemon-app/script.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        .then(res => {
+          if (res.ok) {
+            return res.json()
+          } else {
+            throw `ERROR: ${res.status} ${res.statusText}`
+          }
+        })
+        .then((jsonData) => {
```
Comment (jasperkennis) [inline Week2/homework/pokemon-app/script.js:39]: You are "chaining" here; that is the term when a method is called on the result of a previous result, and so forth.

While you can call `.then().then()`, you don't have to do it here. Can you try to handle the result of `fetch(url)` with just one `.then`?

**Review by jasperkennis — CHANGES_REQUESTED** [review]
Hi Nizami!

Nice to see how much progress you made since the last time I reviewed homework! There are a few things here that need further improvement. Feel free to reach out to me should you have questions.

Cheers! Jasper

**Comment by jasperkennis** [comment]
Hi Nizami, I was just informed that the exercises are optional this week. So you can disregard my comments on them if you wish. But I would recommend to still take a look, I think they will help your understanding of the concepts:)

Great work doing the exercises even though they were optional!

**Comment by nizamimursudlu** [comment]
Hello Jasper. Thanks a lot for helping me out with my homework again. I will definitely pay attantion to all of your comments and try to improve it!

**Comment by nizamimursudlu** [comment]
Hey Jasper, I did change code in exercises #01 and #02.
In exercise #01 I did not use .catch because… setTimeout never fails… please correct me from wrong.

I have a question about the project pokemon-app <https://github.com/HackYourHomework/JavaScript3/pull/16#discussion_r556693225>.



As I understood your comment says that I can use only one .then in stead of two  .then .then

First .then has only headers, so you always have to return json to get the body in the second .then
Second .then is an arrow function…

I didn’t know it is possible to combine them in one .then

Please teach me how

**Comment by jasperkennis** [comment]
Hi Nizami!

Sharp observations😊  

**About exercise 1**
There is really no such thing as code that never fails. Unfortunately, it's often exactly the bit of code that you would trust to always work that bites you. 

Some background (you can skip this if you're in a hurry): When you start building applications for real world production environments, that have to stay active very long or have to handle large amounts of data, you will see failures occurring for example because a machine runs out of memory, or suddenly loses connection to a database, etc etc etc. Without catch, this kind of thing will often cause the entire process to restart, potentially causing data loss. I've actually had many such problems over the past year, we were building a new application that handles orders from Shopify.

TLDR: it's a good idea to get in the habit of always adding catch, even if the action is relatively simple.

Also, currently you have this:

```js
reject(console.log("You didn't pass in a first name!"))
```

but try to change it to:

```js
reject("You didn't pass in a first name!")
```

and try calling the function without passing the name argument, and without adding catch; you will see that it crashes. In many situations, you will actually intentionally use "reject", or even throw an error on purpose, and assume this error will be handled correctly by whoever uses your method.

**About the .thens**
I see many students using this double then, it's not something I personally use or see a lot in my work but it seems that this is much more normal than I thought.

Still, just to for context, here is how I would write that part:

```js
fetch(url)
  .then(res => {
    if (!res.ok) { // Notice negative check: if NOT okay, throw an error.
      throw `ERROR: ${res.status} ${res.statusText}`
    } // else not required because throwing error will stop execution of function, so everything after the error will never happen

    const jsonData = res.json()
    // all the other stuff from the second .then
  });

**Comment by nizamimursudlu** [comment]
Thank a lot Jasper!

**Comment by jasperkennis** [comment]
Running ahead of what you mentioned in the call, I'd made an assumption that res.json() would be synchronous, but it's a promise. Therefor my whole "simple" solution isn't that simpel at all, and I might just run the risk of making things more confusing for you.

But since we have gone down this path now anyway, let me at least show the correct way that I would simplify this:

```js
fetch(url)
  // See ASYNC here:
  .then(async (res) => {
    if (!res.ok) { // Notice negative check: if NOT okay, throw an error.
      throw `ERROR: ${res.status} ${res.statusText}`
    } // else not required because throwing error will stop execution of function, so everything after the error will never happen

    // This becomes an await: 
    const jsonData = await res.json()
  });
```

I tested this, like I should have the first time, this works😊

**Comment by nizamimursudlu** [comment]
Hey Jasper,  I rewrote both functions in pokemon-app the way you recommend, and did some changes is ex01 and ex 02.

Could you please have a look.

**Comment by jasperkennis** [comment]
Hi Nizami, great, I'm glad you made those changes, now I can really see you got the concept of resolve and reject, and the different purposes of .then and .catch.

Great work!
